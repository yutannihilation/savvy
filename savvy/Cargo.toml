[package]
name = "savvy"
description = "An unfriendly R extension interface"
version.workspace = true
edition.workspace = true
authors.workspace = true
license.workspace = true
repository.workspace = true
readme = "README.md"

[dependencies]
rlang-ffi-lite = { verion = "*", path = "../rlang-ffi-lite" }
once_cell = "1"

savvy-macro = { version = "0", path = "../savvy-macro" }

[build-dependencies]
cc = "1"

[package.metadata.dist]
dist = false
