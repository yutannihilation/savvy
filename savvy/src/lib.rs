//! # Savvy - An Unfriendly R Interface
//!

#![doc = include_str!("../docs/design.md")]

pub mod error;
pub mod protect;
pub mod sexp;
pub mod unwind_protect;

pub use error::{Error, Result};
pub use sexp::integer::{IntegerSxp, OwnedIntegerSxp};
pub use sexp::list::{ListElement, ListSxp, OwnedListSxp};
pub use sexp::logical::{LogicalSxp, OwnedLogicalSxp};
pub use sexp::null::NullSxp;
pub use sexp::real::{OwnedRealSxp, RealSxp};
pub use sexp::string::{OwnedStringSxp, StringSxp};
pub use sexp::Sxp;

pub use sexp::external_pointer::{get_external_pointer_addr, IntoExtPtrSxp};

pub use unwind_protect::unwind_protect;

// re-export
pub use rlang_ffi_lite::SEXP;
pub use savvy_macro::savvy;

use rlang_ffi_lite::{cetype_t_CE_UTF8, REprintf, Rf_allocVector, Rf_mkCharLenCE, Rprintf};

use std::ffi::CString;

// TODO: make this r_println! macro
pub fn r_print(msg: &str) -> crate::error::Result<SEXP> {
    unsafe {
        let msg_c_string = CString::new(msg).unwrap();
        unwind_protect(|| {
            Rprintf(msg_c_string.as_ptr());
            NullSxp.into()
        })
    }
}

pub fn r_eprint(msg: &str) -> crate::error::Result<SEXP> {
    unsafe {
        let msg_c_string = CString::new(msg).unwrap();
        unwind_protect(|| {
            REprintf(msg_c_string.as_ptr());
            NullSxp.into()
        })
    }
}

fn alloc_vector(arg1: u32, arg2: isize) -> crate::error::Result<SEXP> {
    unsafe { unwind_protect(|| Rf_allocVector(arg1, arg2)) }
}

// This wrapper function handles Error and panicks, and flag it by setting the
// lowest bit to 1. The lowest bit is supposed to be detected (and then removed)
// on the corresponding C function.
//
// cf. https://en.wikipedia.org/wiki/Tagged_pointer
pub fn handle_error(e: crate::error::Error) -> SEXP {
    match e {
        // The token is already tagged, so pass it as it is.
        error::Error::Aborted(token) => token,

        // In other cases, return the error string with the tag
        e => unsafe {
            let msg = e.to_string();
            let r_error = Rf_mkCharLenCE(
                msg.as_ptr() as *const i8,
                msg.len() as i32,
                cetype_t_CE_UTF8,
            );

            // set the error flag
            (r_error as usize | 1) as SEXP
        },
    }
}
