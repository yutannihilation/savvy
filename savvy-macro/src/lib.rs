use proc_macro::TokenStream;
use quote::quote;

use savvy_bindgen::{SavvyFn, SavvyImpl};

#[proc_macro_attribute]
pub fn savvy(_args: TokenStream, input: TokenStream) -> TokenStream {
    if let Ok(item_fn) = syn::parse::<syn::ItemFn>(input.clone()) {
        match savvy_fn(&item_fn) {
            Ok(result) => return result,
            Err(e) => return e.into_compile_error().into(),
        }
    }

    let parse_result = syn::parse::<syn::ItemImpl>(input.clone());
    if let Ok(item_impl) = parse_result {
        match savvy_impl(&item_impl) {
            Ok(result) => return result,
            Err(e) => return e.into_compile_error().into(),
        }
    }

    proc_macro::TokenStream::from(
        syn::Error::new(
            parse_result.unwrap_err().span(),
            "#[savvy] macro only accepts `Fn` or `Impl`",
        )
        .into_compile_error(),
    )
}

fn savvy_fn(item_fn: &syn::ItemFn) -> syn::Result<TokenStream> {
    let savvy_fn = SavvyFn::from_fn(item_fn)?;

    let item_fn_inner = savvy_fn.generate_inner_fn();
    let item_fn_outer = savvy_fn.generate_outer_fn();

    Ok(quote! {
        #item_fn_inner
        #item_fn_outer
    }
    .into())
}

fn savvy_impl(item_impl: &syn::ItemImpl) -> syn::Result<TokenStream> {
    let savvy_impl = SavvyImpl::new(item_impl)?;
    let orig = savvy_impl.orig.clone();
    let ty = savvy_impl.ty.clone();

    let list_fn_inner = savvy_impl.generate_inner_fns();
    let list_fn_outer = savvy_impl.generate_outer_fns();

    Ok(quote! {
        #orig

        impl savvy::IntoExtPtrSexp for #ty {}

        impl TryFrom<#ty> for savvy::Sexp {
            type Error = savvy::Error;

            fn try_from(value: #ty) -> savvy::Result<Self> {
                use savvy::IntoExtPtrSexp;

                Ok(value.into_external_pointer())
            }
        }

        impl TryFrom<savvy::Sexp> for &mut #ty {
            type Error = savvy::Error;

            fn try_from(value: savvy::Sexp) -> savvy::Result<Self> {
                // Return error if the SEXP is not an external pointer
                value.assert_external_pointer()?;

                let x = unsafe { savvy::get_external_pointer_addr(value.0)? as *mut #ty };
                let res = unsafe { x.as_mut() };
                res.ok_or("Failed to convert the external pointer to the Rust object".into())
            }
        }

        impl TryFrom<savvy::Sexp> for &#ty {
            type Error = savvy::Error;

            fn try_from(value: savvy::Sexp) -> savvy::Result<Self> {
                // Return error if the SEXP is not an external pointer
                value.assert_external_pointer()?;

                let x = unsafe { savvy::get_external_pointer_addr(value.0)? as *mut #ty };
                let res = unsafe { x.as_ref() };
                res.ok_or("Failed to convert the external pointer to the Rust object".into())
            }
        }

        impl TryFrom<savvy::Sexp> for #ty {
            type Error = savvy::Error;

            fn try_from(value: savvy::Sexp) -> savvy::Result<Self> {
                // Return error if the SEXP is not an external pointer
                value.assert_external_pointer()?;

                unsafe { savvy::take_external_pointer_value::<#ty>(value.0) }
            }
        }

        #(#list_fn_inner)*
        #(#list_fn_outer)*
    }
    .into())
}

#[cfg(test)]
mod tests {
    use super::*;
    use syn::parse_quote;

    fn assert_eq_inner(orig: syn::ItemFn, expected: syn::ItemFn) {
        let result = SavvyFn::from_fn(&orig)
            .expect("Failed to parse a function")
            .generate_inner_fn();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_generate_inner_fn() {
        assert_eq_inner(
            parse_quote!(
                #[savvy]
                fn foo() -> savvy::Result<savvy::Sexp> {
                    bar()
                }
            ),
            parse_quote!(
                unsafe fn savvy_foo_inner() -> savvy::Result<savvy::Sexp> {
                    bar()
                }
            ),
        );

        assert_eq_inner(
            parse_quote!(
                #[savvy]
                fn foo() -> savvy::Result<()> {
                    bar();
                    Ok(())
                }
            ),
            parse_quote!(
                unsafe fn savvy_foo_inner() -> savvy::Result<()> {
                    bar();
                    Ok(())
                }
            ),
        );

        #[rustfmt::skip]
        assert_eq_inner(
            // The qualified form (with `savvy::`) and non-qualified form is
            // kept between conversions.
            parse_quote!(
                #[savvy]
                fn foo(x: RealSexp, y: savvy::IntegerSexp) -> savvy::Result<savvy::Sexp> {
                    bar()
                }
            ),
            parse_quote!(
                unsafe fn savvy_foo_inner(x: savvy::ffi::SEXP, y: savvy::ffi::SEXP) -> savvy::Result<savvy::Sexp> {
                    let x = <RealSexp>::try_from(savvy::Sexp(x))?;
                    let y = <savvy::IntegerSexp>::try_from(savvy::Sexp(y))?;
                    bar()
                }
            ),
        );

        #[rustfmt::skip]
        assert_eq_inner(
            parse_quote!(
                #[savvy]
                fn foo(x: f64) -> savvy::Result<savvy::Sexp> {
                    bar()
                }
            ),
            parse_quote!(
                unsafe fn savvy_foo_inner(x: savvy::ffi::SEXP) -> savvy::Result<savvy::Sexp> {
                    let x = <f64>::try_from(savvy::Sexp(x))?;
                    bar()
                }
            ),
        );
    }

    fn assert_eq_outer(orig: syn::ItemFn, expected: syn::ItemFn) {
        let result = SavvyFn::from_fn(&orig)
            .expect("Failed to parse an impl")
            .generate_outer_fn();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_generate_outer_fn() {
        assert_eq_outer(
            parse_quote!(
                #[savvy]
                fn foo() -> savvy::Result<savvy::Sexp> {
                    bar()
                }
            ),
            parse_quote!(
                #[allow(clippy::missing_safety_doc)]
                #[no_mangle]
                pub unsafe extern "C" fn foo() -> savvy::ffi::SEXP {
                    match savvy_foo_inner() {
                        Ok(result) => result.0,
                        Err(e) => savvy::handle_error(e),
                    }
                }
            ),
        );

        assert_eq_outer(
            parse_quote!(
                #[savvy]
                fn foo() -> savvy::Result<()> {
                    bar();
                    Ok(())
                }
            ),
            parse_quote!(
                #[allow(clippy::missing_safety_doc)]
                #[no_mangle]
                pub unsafe extern "C" fn foo() -> savvy::ffi::SEXP {
                    match savvy_foo_inner() {
                        Ok(_) => savvy::sexp::null::null(),
                        Err(e) => savvy::handle_error(e),
                    }
                }
            ),
        );

        #[rustfmt::skip]
        assert_eq_outer(
            parse_quote!(
                #[savvy]
                fn foo(
                    x: RealSexp,
                    y: savvy::RealSexp,
                ) -> savvy::Result<savvy::Sexp> {
                    bar()
                }
            ),
            parse_quote!(
                #[allow(clippy::missing_safety_doc)]
                #[no_mangle]
                pub unsafe extern "C" fn foo(
                    x: savvy::ffi::SEXP,
                    y: savvy::ffi::SEXP
                ) -> savvy::ffi::SEXP {
                    match savvy_foo_inner(x, y) {
                        Ok(result) => result.0,
                        Err(e) => savvy::handle_error(e),
                    }
                }
            ),
        );
    }

    fn assert_eq_outer_impl(orig: &syn::ItemImpl, expected: syn::ItemFn, i: usize) {
        let result =
            SavvyImpl::new(orig).expect("Failed to parse an impl").fns[i].generate_outer_fn();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_generate_outer_fn_impl() {
        let impl1: syn::ItemImpl = parse_quote!(
            #[savvy]
            impl Person {
                fn new() -> Self {
                    Self {}
                }
                fn name(&self) -> savvy::Result<savvy::Sexp> {
                    Ok(out.into())
                }
                fn set_name(&self, name: StringSexp) -> Result<()> {
                    Ok(())
                }
            }
        );

        #[rustfmt::skip]
        assert_eq_outer_impl(
            &impl1,
            parse_quote!(
                #[allow(clippy::missing_safety_doc)]
                #[no_mangle]
                pub unsafe extern "C" fn Person_new() -> savvy::ffi::SEXP {
                    match savvy_Person_new_inner() {
                        Ok(result) => {
                            use savvy::IntoExtPtrSexp;
                            result.into_external_pointer().0
                        },
                        Err(e) => savvy::handle_error(e),
                    }
                }
            ),
            0,
        );

        // TODO
        //
        // assert_eq_outer_impl(
        //     &impl1,
        //     parse_quote!(
        //         #[allow(clippy::missing_safety_doc)]
        //         #[no_mangle]
        //         pub unsafe extern "C" fn Person_name(self__: savvy::ffi::SEXP) -> savvy::ffi::SEXP {
        //             match savvy_Person_name_inner(self__) {
        //                 Ok(result) => result,
        //                 Err(e) => savvy::handle_error(e),
        //             }
        //         }
        //     ),
        //     1,
        // );

        #[rustfmt::skip]
        assert_eq_outer_impl(
            &impl1,
            parse_quote!(
                #[allow(clippy::missing_safety_doc)]
                #[no_mangle]
                pub unsafe extern "C" fn Person_set_name(
                    self__: savvy::ffi::SEXP,
                    name: savvy::ffi::SEXP
                ) -> savvy::ffi::SEXP {
                    match savvy_Person_set_name_inner(self__, name) {
                        Ok(_) => savvy::sexp::null::null(),
                        Err(e) => savvy::handle_error(e),
                    }
                }
            ),
            2,
        );
    }
}
