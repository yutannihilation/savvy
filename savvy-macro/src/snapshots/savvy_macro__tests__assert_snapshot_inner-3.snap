---
source: savvy-macro/src/lib.rs
expression: lines
---
- unsafe fn savvy_foo_inner(
- "    x: savvy::ffi::SEXP,"
- "    y: savvy::ffi::SEXP,"
- ") -> savvy::Result<savvy::Sexp> {"
- "    let orig_hook = std::panic::take_hook();"
- "    std::panic::set_hook(Box::new(savvy::panic_hook::panic_hook));"
- "    let result = std::panic::catch_unwind(|| {"
- "        let x = <RealSexp>::try_from(savvy::Sexp(x)).map_err(|e| e.with_arg_name(\"x\"))?;"
- "        let y = <savvy::IntegerSexp>::try_from(savvy::Sexp(y))"
- "            .map_err(|e| e.with_arg_name(\"y\"))?;"
- "        foo(x, y)"
- "    });"
- "    std::panic::set_hook(orig_hook);"
- "    match result {"
- "        Ok(orig_result) => orig_result,"
- "        Err(_) => Err(\"panic happened\".into()),"
- "    }"
- "}"
